### PR TITLE
Improve reflection system and static_variant

### PR DESCRIPTION
**Pull request by nathanielhourt**
Previously, fc reflection generated visitor infrastructure capable of
visiting each field of a struct; however, this is only useful for run-
time operations on structures. No compile-time information was preserved
about the fields of a struct, making compile-time introspection of types
impossible.

Now FC reflection generates rich compile-time information about the
members of structs, including types, names, and structure, allowing
code to be written to explore types in detail at compile-time and fetch
arbitrary fields at runtime without needing to iterate over the unwanted
fields.

To make this a reality, a new `typelist` type was added to store this
compile-time information. This type is also useful within the context of
`static_variant`, as SV previously used implementation detail types in a
private namespace to provide equivalent functionality. Now `static_variant`
uses the reusable functionality of `typelist` to operate, dramatically
reducing the amount of code dedicated to `static_variant` and also
making `static_variant` types more flexible to work with since their
infrastructure is now based on the `typelist` public interface.

These changes are necessary for bitshares/bitshares-core#1860

### CODE REVIEW

File: include/fc/reflect/reflect.hpp [hunk 1]
```diff
@@ -257,25 +375,11 @@ namespace fc { \
...
-#define FC_REFLECT_IMPL( TYPE, MEMBERS ) \
-    FC_REFLECT_DERIVED_IMPL_EXT( TYPE, BOOST_PP_SEQ_NIL, MEMBERS )
-
+// vvv These are preserved to merge more smoothly vvv
+#define FC_REFLECT_BASE_MEMBER_COUNT( r, OP, elem ) \
+  OP fc::reflector<elem>::total_member_count
 
+#define FC_REFLECT_MEMBER_COUNT( r, OP, elem ) \
+  OP 1
+// ^^^ These can be removed after updating graphene/protocol/types.hpp ^^^
```
Comment (abitmore) [inline include/fc/reflect/reflect.hpp]: I think it's better to just remove them, most likely we'll bump FC and update `types.hpp` in the same PR.

File: include/fc/static_variant.hpp [hunk 2]
```diff
@@ -400,36 +260,40 @@ class static_variant {
...
-        static const auto wrappers = init_const_wrappers<impl::type_info<Types...>::count,const visitor,const void*,Types...>( v, data );
         FC_ASSERT( tag >= 0 && tag < count(), "Unsupported type ${tag}!", ("tag",tag) );
-        return wrappers[tag]( v, data );
+        return typelist::runtime::dispatch(list(), tag, [&v, data](auto t) {
+            return v(*reinterpret_cast<const typename decltype(t)::type*>(data));
+        });
     }
 
-    static constexpr int count() { return impl::type_info<Types...>::count; }
+    static int count() { return typelist::length<list>(); }
```
Comment (abitmore) [inline include/fc/static_variant.hpp]: Since typelist::length() is still `constexpr`, why `count()` is no longer `constexpr`?

File: include/fc/static_variant.hpp [hunk 3]
```diff
@@ -280,28 +77,69 @@ class static_variant {
...
-    template<typename StaticVariant>
-    friend struct impl::copy_construct;
-    template<typename StaticVariant>
-    friend struct impl::move_construct;
 public:
     template<typename X, typename = type_in_typelist<X>>
     struct tag
     {
-       static constexpr int value = impl::position<X, Types...>::pos;
+       static const int value = typelist::index_of<list, X>();
```
Comment (abitmore) [inline include/fc/static_variant.hpp]: Why not `constexpr`?

File: include/fc/static_variant.hpp (hunk 2 above)
Comment (nathanielhourt) [reply include/fc/static_variant.hpp]: Oh haha Peter fixed those to be constexpr, but that was after I forked, and it got lost in the rebase. I'll fix it.

File: include/fc/reflect/reflect.hpp (hunk 1 above)
Comment (nathanielhourt) [reply include/fc/reflect/reflect.hpp]: Meh, OK

File: src/network/http/websocket.cpp [hunk 4]
```diff
@@ -300,6 +300,10 @@ namespace fc { namespace http {
                   _server.close( item.first, 0, "server exit" );
 
                if( _closed ) _closed->wait();
+#ifdef TRAVIS_BUILD
```
Comment (pmconrad) [inline src/network/http/websocket.cpp]: I thought abit had fixed this somewhere else?
Is it still required? If it is should be in a separate PR.

Comment (nathanielhourt) [reply src/network/http/websocket.cpp]: Not sure, but I'll remove it.

Comment (oxarbitrage) [reply src/network/http/websocket.cpp]: iirc it was a hack and we never included it. not against adding it.

**Review by pmconrad — DISMISSED** [review]
Ah, you make this look so simple! Excellent work! :-)
@abitmore?

**Comment by abitmore** [comment]
@pmconrad I haven't fixed the `TRAVIS_BUILD` issue yet, it does not only affect Travis though.

**Comment by nathanielhourt** [comment]
> Ah, you make this look so simple! Excellent work! :-)

Thank you very much, Peter =) That means a great deal to me.

**Comment by nathanielhourt** [comment]
Shiny new `static_variant::operator==` in a75fcb56f3cd51eb28d43673c9ea583fe23cfe34

I'm opting to leave out a default `operator<` as it's not immediately obvious how to resolve comparison between unmatched types. Not to mention, our `flat_set` usages depend on the fact that `(static_variant<int>(1) < static_variant<int>(2)) == false`. `type_lt` gets us that, but it's dangerous to add a default `operator<` in light of this fact.

**Comment by pmconrad** [comment]
Please `rebase -i` and fixup the travis workaround removal.

**Comment by nathanielhourt** [comment]
Done

**Review by pmconrad — APPROVED** [review]
The fixup went into the wrong commit, but nevermind.
